File: src/com/cattailsw/nanidroid/GhostPresentationFrame.java
package com.cattailsw.nanidroid;

/**
 * Immutable render input emitted by the Sakura Script runtime.
 *
 * <p>This deliberately contains no Android view types. The legacy renderer is
 * one consumer; a Compose renderer will consume the same presentation facts
 * after the runtime migration.</p>
 */
public final class GhostPresentationFrame {
    public static final class Speaker {
        public final String text;
        public final String surfaceId;
        public final String animationId;
        public final boolean balloonVisible;

        public Speaker(String text, String surfaceId, String animationId, String balloonId) {
            this.text = text;
            this.surfaceId = surfaceId;
            this.animationId = animationId;
            this.balloonVisible = !"-1".equalsIgnoreCase(balloonId) || text.length() > 0;
        }
    }

    public final Speaker sakura;
    public final Speaker kero;
    public final boolean talkingAnimationEnabled;

    public GhostPresentationFrame(
            Speaker sakura, Speaker kero, boolean talkingAnimationEnabled) {
        this.sakura = sakura;
        this.kero = kero;
        this.talkingAnimationEnabled = talkingAnimationEnabled;
    }
}

File: src/com/cattailsw/nanidroid/GhostPresentationFrame.kt
package com.cattailsw.nanidroid

/**
 * Immutable, UI-free render input emitted by the Sakura Script runtime.
 *
 * The fields remain directly accessible from the legacy Java renderer while
 * Compose consumes the same presentation facts without View types.
 */
data class GhostPresentationFrame(
    @JvmField val sakura: Speaker,
    @JvmField val kero: Speaker,
    @JvmField val talkingAnimationEnabled: Boolean,
) {
    class Speaker(
        @JvmField val text: String,
        @JvmField val surfaceId: String?,
        @JvmField val animationId: String?,
        private val balloonId: String,
    ) {
        @JvmField
        val balloonVisible: Boolean =
            !balloonId.equals("-1", ignoreCase = true) || text.isNotEmpty()
    }
}
